# 2/14  1. Введение

ВСПОМНИМ, С КАКИМИ ДАННЫМИ МЫ РАБОТАЕМ

In [2]:
import pandas as pd
import numpy as np

melb_dfc = pd.read_csv('data/melb_data_fe.csv')
melb_df = melb_dfc.copy()
melb_df.head()

Suburb  Rooms   Type      Price Method SellerG        Date  Distance  \
0  Abbotsford      2  house  1480000.0      S  Biggin  2016-03-12       2.5   
1  Abbotsford      2  house  1035000.0      S  Biggin  2016-04-02       2.5   
2  Abbotsford      3  house  1465000.0     SP  Biggin  2017-04-03       2.5   
3  Abbotsford      3  house   850000.0     PI  Biggin  2017-04-03       2.5   
4  Abbotsford      4  house  1600000.0     VB  Nelson  2016-04-06       2.5   

   Postcode  Bedroom  ...  Longtitude             Regionname  Propertycount  \
0      3067        2  ...    144.9984  Northern Metropolitan           4019   
1      3067        2  ...    144.9934  Northern Metropolitan           4019   
2      3067        3  ...    144.9944  Northern Metropolitan           4019   
3      3067        3  ...    144.9969  Northern Metropolitan           4019   
4      3067        3  ...    144.9941  Northern Metropolitan           4019   

   MeanRoomsSquare AreaRatio  MonthSale  AgeBuilding WeekdaySale  StreetType  \
0            25.20 -0.231707          3           46           5          St   
1            15.80 -0.327660          4          116           5          St   
2            18.75  0.056338          4          117           0          St   
3            15.75  0.145455          4           47           0       other   
4            17.75  0.083969          4            2           2          St   

   Weekend  
0        1  
1        1  
2        0  
3        0  
4        0  

[5 rows x 26 columns]

In [3]:
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Suburb           13580 non-null  object 
 1   Rooms            13580 non-null  int64  
 2   Type             13580 non-null  object 
 3   Price            13580 non-null  float64
 4   Method           13580 non-null  object 
 5   SellerG          13580 non-null  object 
 6   Date             13580 non-null  object 
 7   Distance         13580 non-null  float64
 8   Postcode         13580 non-null  int64  
 9   Bedroom          13580 non-null  int64  
 10  Bathroom         13580 non-null  int64  
 11  Car              13580 non-null  int64  
 12  Landsize         13580 non-null  float64
 13  BuildingArea     13580 non-null  float64
 14  CouncilArea      12211 non-null  object 
 15  Lattitude        13580 non-null  float64
 16  Longtitude       13580 non-null  float64
 17  Regionname  

None

`csv`-файл не хранит в себе информацию о типах данных столбцов, поэтому при чтении Pandas автоматически определяет тип данных столбца. Не забывайте об этом, обмениваясь преобразованными данными с вашими коллегами.

Задание 1.1

Преобразуйте столбец `Date` в формат `datetime` и выделите квартал (`quarter`) продажи объектов недвижимости. Найдите второй по популярности квартал продажи. В качестве ответа запишите число объектов, проданных в этом квартале.

In [4]:
melb_df['Date'] = pd.to_datetime(melb_df['Date'])
melb_df['Date'].dt.quarter.value_counts()


Date
3    4873
2    4359
4    2329
1    2019
Name: count, dtype: int64

Задание 1.2

Преобразуйте все столбцы, в которых меньше 150 уникальных значений, в тип данных `category`, исключив из преобразования столбцы `Date`, `Rooms`, `Bedroom`, `Bathroom`, `Car`.
В качестве ответа запишите результирующее количество столбцов, которые имеют тип данных `category`.

In [5]:
unique_list = []
for col in melb_df.columns:
    item = (col, melb_df[col].nunique(), melb_df[col].dtype)
    unique_list.append(item)
unique_counts = pd.DataFrame(
    unique_list, columns=[
        'Colomns', 'Unique', 'Dtype']).sort_values(
            by='Unique', ignore_index=True)
        
display(unique_counts)


Colomns  Unique           Dtype
0           Weekend       2           int64
1              Type       3          object
2            Method       5          object
3       WeekdaySale       7           int64
4        Regionname       8          object
5             Rooms       9           int64
6          Bathroom       9           int64
7               Car      11           int64
8        StreetType      11          object
9         MonthSale      12           int64
10          Bedroom      12           int64
11      CouncilArea      33          object
12          SellerG      50          object
13             Date      58  datetime64[ns]
14           Suburb     120          object
15      AgeBuilding     152           int64
16         Postcode     198           int64
17         Distance     202         float64
18    Propertycount     311           int64
19     BuildingArea     602         float64
20  MeanRoomsSquare    1275         float64
21         Landsize    1448         float64
22            Price    2204         float64
23        Lattitude    6503         float64
24        AreaRatio    6521         float64
25       Longtitude    7063         float64

In [6]:
cols_to_exclude = ['Date', 'Rooms', 'Bedroom', 'Bathroom', 'Car'] # Список столбцов, которые мы не берём во внимание
max_unique_count = 150 # Задаём максимальное число уникальных категорий
for col in melb_df.columns: # Цикл по именам столбцов
    if melb_df[col].nunique() < max_unique_count and col not in cols_to_exclude: # Проверяем условие
        melb_df[col] = melb_df[col].astype('category') # Преобразуем тип столбца
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Suburb           13580 non-null  category      
 1   Rooms            13580 non-null  int64         
 2   Type             13580 non-null  category      
 3   Price            13580 non-null  float64       
 4   Method           13580 non-null  category      
 5   SellerG          13580 non-null  category      
 6   Date             13580 non-null  datetime64[ns]
 7   Distance         13580 non-null  float64       
 8   Postcode         13580 non-null  int64         
 9   Bedroom          13580 non-null  int64         
 10  Bathroom         13580 non-null  int64         
 11  Car              13580 non-null  int64         
 12  Landsize         13580 non-null  float64       
 13  BuildingArea     13580 non-null  float64       
 14  CouncilArea      12211 non-null  categ

None

# 3/14  2. Сортировка данных в DataFrame

МЕТОД `SORT_VALUES()`

Основные параметры метода `sort_values()`

`by` — имя или список имён столбцов, по значениям которых производится сортировка.  
`axis` — ось, по которой производится сортировка (`0` — строки, `1` — столбцы). По умолчанию сортировка производится по строкам.  
`ascending` — сортировка по **возрастанию** (от меньшего к большему). По умолчанию параметр выставлен на `True`, для сортировки по **убыванию** (от большего к меньшему) необходимо выставить его на `False`.  
`ignore_index` — создаются ли новые индексы в таблице. По умолчанию выставлен на `False` и сохраняет индексы изначальной таблицы.  
`inplace` — производится ли замена исходной таблицы на отсортированную. По умолчанию параметр выставлен на `False`, то есть замены не производится. Чтобы переопределить исходную таблицу на отсортированную, необходимо выставить этот параметр на `True`.

СОРТИРОВКА ПО ЗНАЧЕНИЯМ ОДНОГО СТОЛБЦА

In [20]:
melb_df.sort_values(by='Price').head(10)

Suburb  Rooms   Type     Price Method        SellerG  \
2652        Footscray      1   unit   85000.0     PI        Burnham   
1805            other      4  house  131000.0     PI          other   
7303           Albion      1   unit  145000.0     PI         Biggin   
1927           Coburg      4  house  145000.0     PI         Jellis   
7940         Hawthorn      1   unit  160000.0     VB            HAR   
12666       Brunswick      1   unit  170000.0     VB         Nelson   
8811        Footscray      1   unit  170000.0     PI        Burnham   
8504   West Footscray      1   unit  185000.0     PI            Jas   
7293           Albion      1   unit  185000.0      S  hockingstuart   
7305           Albion      2   unit  190000.0     SP        Burnham   

            Date  Distance  Postcode  Bedroom  ...  Longtitude  \
2652  2016-03-09       6.4      3011        1  ...   144.89000   
1805  2017-02-25       8.9      3162        4  ...   145.02420   
7303  2016-05-28      13.9      3020        2  ...   144.82660   
1927  2016-04-06       7.8      3058        3  ...   144.96580   
7940  2017-08-04       4.6      3122        1  ...   145.03730   
12666 2017-09-16       5.2      3056        1  ...   144.95188   
8811  2017-01-07       5.1      3011        1  ...   144.89587   
8504  2017-04-29       8.2      3012        1  ...   144.86720   
7293  2016-08-22      13.9      3020        1  ...   144.82720   
7305  2016-07-30      13.9      3020        2  ...   144.82390   

                  Regionname  Propertycount  MeanRoomsSquare AreaRatio  \
2652    Western Metropolitan           7570        42.000000  1.000000   
1805   Southern Metropolitan           2379        17.222222 -0.525994   
7303    Western Metropolitan           2185        31.500000  0.555556   
1927   Northern Metropolitan          11204        20.500000 -0.531429   
7940   Southern Metropolitan          11308        42.000000 -0.437500   
12666  Northern Metropolitan          11918        42.000000 -0.816860   
8811    Western Metropolitan           7570         8.666667 -0.071429   
8504    Western Metropolitan           5058        42.000000  1.000000   
7293    Western Metropolitan           2185        14.333333 -0.964017   
7305    Western Metropolitan           2185        25.200000  1.000000   

       MonthSale  AgeBuilding WeekdaySale  StreetType  Weekend  
2652           3            9           2          St        0  
1805           2           97           5          St        1  
7303           5           46           5          St        1  
1927           4          106           2          Rd        0  
7940           8            8           4          St        0  
12666          9           47           5          St        1  
8811           1            4           5          St        1  
8504           4           47           5          St        1  
7293           8           41           0          Rd        0  
7305           7           46           5          St        1  

[10 rows x 26 columns]

In [8]:
melb_df.sort_values(by='Date', ascending=False).head(10)

Suburb  Rooms   Type      Price Method        SellerG  \
11144      Northcote      4  house  1955000.0     SP        McGrath   
11217   Surrey Hills      3  house  1775000.0     PI         Jellis   
11206       St Kilda      4  house  1600000.0     VB           Gary   
11207     Strathmore      7  house  2000000.0     SP         Nelson   
11208     Strathmore      4  house  1610000.0      S          other   
11209     Strathmore      3  house  1742000.0      S          Barry   
11210     Strathmore      3  house  1016000.0      S          other   
11211       Sunshine      3  house   700000.0     SP          other   
11212       Sunshine      2  house   615000.0     SP          Barry   
11213  Sunshine West      3  house   650000.0     VB  hockingstuart   

            Date  Distance  Postcode  Bedroom  ...  Longtitude  \
11144 2017-12-08       5.3      3070        4  ...   144.99375   
11217 2017-12-08      10.2      3127        3  ...   145.08968   
11206 2017-12-08       5.0      3182        4  ...   144.98324   
11207 2017-12-08       8.2      3041        7  ...   144.90587   
11208 2017-12-08       8.2      3041        4  ...   144.92243   
11209 2017-12-08       8.2      3041        3  ...   144.92172   
11210 2017-12-08       8.2      3041        3  ...   144.91521   
11211 2017-12-08      10.5      3020        3  ...   144.84569   
11212 2017-12-08      10.5      3020        2  ...   144.83206   
11213 2017-12-08      10.5      3020        3  ...   144.80099   

                  Regionname  Propertycount  MeanRoomsSquare AreaRatio  \
11144  Northern Metropolitan          11364        12.600000  0.923664   
11217  Southern Metropolitan           5457        15.750000 -0.721854   
11206  Southern Metropolitan          13240        16.000000 -0.586028   
11207   Western Metropolitan           3284        20.882353 -0.262721   
11208   Western Metropolitan           3284        24.909091 -0.517181   
11209   Western Metropolitan           3284        15.750000 -0.693803   
11210   Western Metropolitan           3284        15.750000 -0.413953   
11211   Western Metropolitan           3755        16.250000 -0.577236   
11212   Western Metropolitan           3755        18.800000 -0.632812   
11213   Western Metropolitan           6763        18.000000 -0.691554   

       MonthSale  AgeBuilding WeekdaySale  StreetType  Weekend  
11144         12           47           4          St        0  
11217         12           47           4          Rd        0  
11206         12          107           4          St        0  
11207         12           18           4          Rd        0  
11208         12           62           4          St        0  
11209         12           47           4          Rd        0  
11210         12           47           4          St        0  
11211         12           67           4          St        0  
11212         12           77           4          St        0  
11213         12           47           4          St        0  

[10 rows x 26 columns]

СОРТИРОВКА ПО ЗНАЧЕНИЯМ НЕСКОЛЬКИХ СТОЛБЦОВ

Для сортировки по значениям нескольких столбцов необходимо передать названия этих столбцов в параметр `by` в виде списка. При этом важно обращать внимание на порядок следования столбцов.

Так, например, отсортируем таблицу сначала по возрастанию расстояния от центра города (`Distance`), а затем — по возрастанию цены объекта (`Price`). Для того чтобы вывод был более наглядным, выделим каждую десятую строку из столбцов `Distance` и `Price` результирующей таблицы:

In [9]:
melb_df.sort_values(
    by=['Distance', 'Price']).loc[::10, ['Distance', 'Price']]

Distance      Price
11428       0.0   387000.0
10512       0.7   600000.0
5727        1.2   485000.0
8671        1.2   595000.0
5736        1.2   740000.0
...         ...        ...
12011      38.0   680000.0
10673      38.0   810000.0
13429      38.0  1155000.0
11102      41.0   650000.0
12728      45.9   650000.0

[1358 rows x 2 columns]

Мы получили таблицу, отсортированную по возрастанию расстояния до центра города. Если встречаются объекты недвижимости, у которых расстояние оказывается одинаковым, то внутри такой группы производится сортировка по цене объекта.

Ради интереса попробуйте поменять порядок следования столбцов в параметре `by` метода `sort_values()` и сравните результат. 

In [10]:
melb_df.sort_values(
    by=['Price', 'Distance']).loc[::10, ['Distance', 'Price']]

Distance      Price
2652        6.4    85000.0
7116        7.8   200000.0
5206       11.2   220000.0
7240        8.2   241000.0
2898       13.0   250000.0
...         ...        ...
13341       6.3  4060000.0
8591        1.5  4275000.0
1094       11.2  4700000.0
5801        6.1  5046000.0
251         6.3  5525000.0

[1358 rows x 2 columns]

In [11]:
ddd = {'counts': [np.random.randint(20) for i in range(20)], 'persent':[np.random.randint(20) for i in range(20)]}
ddd1 = pd.DataFrame(ddd)
ddd1.loc[2:19:5,'persent']# Со 2-го по 19-й, но каждый 5-й.

2     10
7      0
12    18
17    19
Name: persent, dtype: int64

КОМБИНИРОВАНИЕ СОРТИРОВКИ С ФИЛЬТРАЦИЕЙ

Найдём информацию о таунхаусах (`Type`), проданных компанией (`SellerG`) `McGrath`, у которых коэффициент соотношения площадей здания и участка (`AreaRatio`) меньше `-0.8`. Результат отсортируем по дате продажи (`Date`) в порядке возрастания, а после проведём сортировку по убыванию коэффициента соотношения площадей. Также обновим старые индексы на новые, установив параметр `ignore_index` на `True`. Для наглядности результата выберем из таблицы только столбцы `Data` и `AreaRatio`:

May option

In [52]:
melb_df[(
    melb_df['Type'] == 'townhouse') & (
        melb_df['SellerG'] == 'McGrath') & (
            melb_df['AreaRatio'] <= -0.8)].sort_values(
                by=['Date', 'AreaRatio'], ascending=[
                    True, False], ignore_index=True).loc[
                        ::, ['Date', 'AreaRatio']]

Date  AreaRatio
0 2016-07-26  -0.974922
1 2016-09-24  -0.971831
2 2016-11-27  -0.953608
3 2016-12-11  -0.945946
4 2017-08-04  -0.947368
5 2017-08-04  -0.970874

Option SkillFactory

In [53]:
mask1 = melb_df['AreaRatio'] < -0.8
mask2 = melb_df['Type'] == 'townhouse'
mask3 = melb_df['SellerG'] == 'McGrath'
melb_df[mask1 & mask2 & mask3].sort_values(
    by=['Date', 'AreaRatio'],
    ascending=[True, False],
    ignore_index=True
).loc[:, ['Date', 'AreaRatio']]

Date  AreaRatio
0 2016-07-26  -0.974922
1 2016-09-24  -0.971831
2 2016-11-27  -0.953608
3 2016-12-11  -0.945946
4 2017-08-04  -0.947368
5 2017-08-04  -0.970874

ПРО ВАЖНОСТЬ КОПИРОВАНИЯ ПРИ СОРТИРОВКЕ С `inplace=Tru`e

Примечание. Старайтесь не сочетать фильтрацию и метод sort_values() с параметром inplace=True, так как в таком случае у вас возникнет предупреждение (warning) SettingWithCopyWarning: melb_df[melb_df['Rooms'] > 5].sort_values(inplace=True, by=['Rooms']):

C:\Users\Andrey\anaconda3\lib\site-packages\pandas\util\_decorators.py:311: SettingwithCopyWarning:

A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/ indexing. html#retu rning-a-view-versus-a-copy

    return func(*args, **kwargs)
    
Стоит отметить, что это не ошибка и код в таком случае отработает. Однако Pandas предупреждает вас о том, что при использовании такого кода дальнейшие результаты могут быть неожиданными.

Это предупреждение предназначено для обозначения операций «цепного присваивания». Это ситуация, когда вы пытаетесь напрямую изменить подмножество исходных данных. В нашем случае мы пытаемся отсортировать данные с заменой (об этом говорит параметр inplace=True) исходной таблицы на отсортированную.

Чтобы не возникало подобных конфликтов, необходимо использовать метод copy() для явного создания копии отфильтрованного подмножества исходных данных и работать уже с ней (производить сортировку):

filtered = melb_df[melb_df['Rooms'] > 5].copy()

filtered.sort_values(inplace=True, by=['Rooms'])

filtered.head()

Задание 2.2

Произведите сортировку столбца `AreaRatio` по убыванию. При этом индексы полученной таблицы замените на новые. Какое значение площади здания находится в строке `1558`? Ответ округлите до целого числа.

In [64]:
melb_df.sort_values(
    by='AreaRatio', ascending=False, ignore_index=True).loc[
        1558, 'BuildingArea']

126.0

Задание 2.3

Найдите таунхаусы (`Type`) с количеством жилых комнат (`Rooms`) больше `2`. Отсортируйте полученную таблицу сначала по возрастанию числа комнат, а затем по убыванию средней площади комнат (`MeanRoomsSquare`). Индексы таблицы замените на новые. Какая цена будет у объекта в строке 18? Ответ запишите в виде целого числа.

In [67]:
melb_df[(melb_df['Type'] == 'townhouse') & (melb_df[
    'Rooms'] > 2)].sort_values(by=[
        'Rooms', 'MeanRoomsSquare'], ascending=[
            True, False], ignore_index=True).loc[18, 'Price']

1300000.0

# 4/14  3. Группировка данных в DataFrame

МЕТОД `GROUPBY()`

В библиотеке Pandas для группировки данных по одному или нескольким признакам можно использовать метод `groupby()`.

Основные параметры метода `groupby()`

`by` — имя или список имён столбцов, по которым производится группировка.
`axis` — ось, по которой производится группировка (`0` — строки, `1` — столбцы). По умолчанию группировка производится по строкам.
`as_index` — добавляется ли дополнительный индекс к таблице. По умолчанию установлен на `True`.

Метод `groupby()` возвращает объект `DataFrameGroupBy`, который хранит в себе информацию о том, какие строки относятся к определённой группе, и сам по себе не представляет для нас интереса. 

Однако к этому объекту можно применять уже знакомые нам агрегирующие методы (`mean`, `median`, `sum` и т. д.), чтобы рассчитывать показатели внутри каждой группы.

ГРУППИРОВКА ДАННЫХ ПО ОДНОМУ КРИТЕРИЮ С ОДНОЙ АГРЕГАЦИЕЙ

In [83]:
melb_df.groupby(by='Type').mean(numeric_only=True)

Rooms         Price   Distance     Postcode   Bedroom  Bathroom  \
Type                                                                            
house      3.260874  1.242665e+06  10.979479  3104.080643  3.229336  1.613822   
townhouse  2.837522  9.337351e+05   9.851346  3100.777379  2.814183  1.809695   
unit       1.963871  6.051275e+05   7.607391  3110.797481  1.966523  1.183295   

                Car    Landsize  BuildingArea  Lattitude  Longtitude  \
Type                                                                   
house      1.772674  617.181924    152.162553 -37.803795  144.994700   
townhouse  1.555655  279.606822    134.649710 -37.815782  144.996489   
unit       1.128936  477.314219    102.235863 -37.823710  144.996363   

           Propertycount  MeanRoomsSquare  AreaRatio  AgeBuilding  
Type                                                               
house        7259.025505        18.996731  -0.490031    55.669700  
townhouse    7094.459605        18.569847  -0.094916    26.690305  
unit         8199.280080        21.068242   0.319883    39.703016

Мы получили таблицу, на пересечении строк и столбцов которой находятся средние значения каждого числового признака в наших данных.

**Примечание**. Если мы хотим видеть тип объекта в качестве отдельного столбца таблицы, мы можем выставить параметр `as_index` на `False`:

In [87]:
melb_df.groupby(by='Type', as_index=False).mean(numeric_only=True)

Type     Rooms         Price   Distance     Postcode   Bedroom  \
0      house  3.260874  1.242665e+06  10.979479  3104.080643  3.229336   
1  townhouse  2.837522  9.337351e+05   9.851346  3100.777379  2.814183   
2       unit  1.963871  6.051275e+05   7.607391  3110.797481  1.966523   

   Bathroom       Car    Landsize  BuildingArea  Lattitude  Longtitude  \
0  1.613822  1.772674  617.181924    152.162553 -37.803795  144.994700   
1  1.809695  1.555655  279.606822    134.649710 -37.815782  144.996489   
2  1.183295  1.128936  477.314219    102.235863 -37.823710  144.996363   

   Propertycount  MeanRoomsSquare  AreaRatio  AgeBuilding  
0    7259.025505        18.996731  -0.490031    55.669700  
1    7094.459605        18.569847  -0.094916    26.690305  
2    8199.280080        21.068242   0.319883    39.703016

Как правило, нам не нужна информация обо всех столбцах, поэтому агрегирующие методы можно применять только к интересующему нас столбцу. Например, давайте сравним средние цены на объекты в зависимости от их типа:

In [82]:
melb_df.groupby('Type')['Price'].mean()

Type
house        1.242665e+06
townhouse    9.337351e+05
unit         6.051275e+05
Name: Price, dtype: float64

**Примечание**. Обратите внимание, что, так как мы считаем только один показатель (среднее) для одного столбца, в результате мы получаем объект `Series`.

Теперь давайте выясним, какие регионы (`Regionname`) наиболее удалены от центра Мельбурна.

In [88]:
melb_df.groupby('Regionname')['Distance'].min().sort_values(ascending=False)

Regionname
Western Victoria              29.8
Eastern Victoria              25.2
Northern Victoria             21.8
South-Eastern Metropolitan    14.7
Eastern Metropolitan           7.8
Western Metropolitan           4.3
Southern Metropolitan          0.7
Northern Metropolitan          0.0
Name: Distance, dtype: float64

ГРУППИРОВКА ДАННЫХ ПО ОДНОМУ КРИТЕРИЮ С НЕСКОЛЬКИМИ АГРЕГАЦИЯМИ

Чтобы рассчитать несколько агрегирующих методов, можно воспользоваться методом `agg()`, который принимает список строк с названиями агрегаций.

Давайте построим таблицу для анализа продаж по месяцам. Для этого найдём количество продаж, а также среднее и максимальное значения цен объектов недвижимости (`Price`), сгруппированных по номеру месяца продажи (`MonthSale`). Результат отсортируем по количеству продаж в порядке убывания:

In [92]:
melb_df.groupby('MonthSale')['Price'].agg(['count', 'max', 'mean']).sort_values(by='count', ascending=False)

count        max          mean
MonthSale                                
8           1850  6500000.0  1.056371e+06
7           1835  9000000.0  9.314698e+05
5           1644  8000000.0  1.097807e+06
6           1469  7650000.0  1.068981e+06
3           1408  5600000.0  1.146762e+06
4           1246  5500000.0  1.050479e+06
9           1188  6400000.0  1.126349e+06
10           854  6250000.0  1.135970e+06
11           750  5050000.0  1.142503e+06
12           725  5700000.0  1.144737e+06
2            333  4735000.0  1.169051e+06
1            278  5200000.0  9.397921e+05

**Примечание**. Обратите внимание, что, так как мы считаем несколько показателей для одного столбца, в результате мы получаем объект `DataFrame`.

**Примечание**. Если вам нужна полная информация обо всех основных статистических характеристиках внутри каждой группы, вы можете воспользоваться методом `agg()`, передав в качестве его параметра строку '`describe`':

In [94]:
melb_df.groupby('MonthSale')['Price'].agg('describe')# = ... .describe()

count          mean            std       min       25%        50%  \
MonthSale                                                                       
1           278.0  9.397921e+05  577668.924214  170000.0  570500.0   795000.0   
2           333.0  1.169051e+06  671564.357417  131000.0  710000.0  1020000.0   
3          1408.0  1.146762e+06  709573.596867   85000.0  680000.0   945000.0   
4          1246.0  1.050479e+06  591892.902979  145000.0  655000.0   905500.0   
5          1644.0  1.097807e+06  668492.867996  145000.0  650000.0   905000.0   
6          1469.0  1.068981e+06  606010.069052  222000.0  660000.0   900000.0   
7          1835.0  9.314698e+05  537390.803161  190000.0  586750.0   800000.0   
8          1850.0  1.056371e+06  619617.476541  160000.0  635000.0   892000.0   
9          1188.0  1.126349e+06  608734.690742  170000.0  725000.0   980000.0   
10          854.0  1.135970e+06  692950.251627  250000.0  652625.0   950000.0   
11          750.0  1.142503e+06  703666.227892  240000.0  671750.0   950000.0   
12          725.0  1.144737e+06  706549.007161  250000.0  667500.0   938000.0   

                 75%        max  
MonthSale                        
1          1111250.0  5200000.0  
2          1478000.0  4735000.0  
3          1400000.0  5600000.0  
4          1298750.0  5500000.0  
5          1371250.0  8000000.0  
6          1325000.0  7650000.0  
7          1150000.0  9000000.0  
8          1310000.0  6500000.0  
9          1360000.0  6400000.0  
10         1416500.0  6250000.0  
11         1399125.0  5050000.0  
12         1412000.0  5700000.0

После базовых математических функций наиболее частым агрегированием является подсчёт числа уникальных значений. Так, например, мы можем вычислить число уникальных риелторских компаний в зависимости от региона, чтобы понять, в каких регионах конкуренция на рынке недвижимости меньше. Это можно сделать, передав в параметр метода `agg()` строку '`nunique`'. 

Более того, метод `agg()` поддерживает использование и других функций. Передадим дополнительно встроенную функцию `set`, чтобы получить множество из агентств недвижимости, которые работают в каждом из регионов:

In [98]:
melb_df.groupby('Regionname')['SellerG'].agg(['nunique', set]).sort_values(by='nunique', ascending=False)

nunique  \
Regionname                            
Northern Metropolitan            40   
Southern Metropolitan            38   
Western Metropolitan             34   
Eastern Metropolitan             26   
South-Eastern Metropolitan       25   
Eastern Victoria                 11   
Northern Victoria                11   
Western Victoria                  6   

                                                                          set  
Regionname                                                                     
Northern Metropolitan       {Jas, Burnham, Philip, Cayzer, Collins, Greg, ...  
Southern Metropolitan       {Jas, Cayzer, Chisholm, McGrath, Greg, Marshal...  
Western Metropolitan        {Jas, Burnham, Chisholm, McGrath, Greg, Barry,...  
Eastern Metropolitan        {Philip, McGrath, Marshall, Barry, Stockdale, ...  
South-Eastern Metropolitan  {Chisholm, McGrath, Greg, Barry, Stockdale, Ne...  
Eastern Victoria            {Eview, McGrath, Fletchers, HAR, C21, other, B...  
Northern Victoria           {Raine, McGrath, HAR, other, hockingstuart, Ba...  
Western Victoria                 {Raine, HAR, other, hockingstuart, Ray, YPA}

Задание 3.1

Сгруппируйте данные по признаку количества комнат и найдите среднюю цену объектов недвижимости в каждой группе. В качестве ответа запишите количество комнат, для которых средняя цена наибольшая.

In [104]:
melb_df.groupby('Rooms')['Price'].mean().sort_values(ascending=False)

Rooms
7     1.920700e+06
5     1.870260e+06
6     1.849366e+06
8     1.602750e+06
4     1.445282e+06
3     1.076081e+06
10    9.000000e+05
2     7.750812e+05
1     4.338245e+05
Name: Price, dtype: float64

In [ ]:
ddd = 2

Задание 3.2

Какой регион имеет наименьшее стандартное отклонение по географической широте (`Lattitude`)?
В качестве ответа запишите название этого региона.